File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        """

        _ _ _ _
        ^ ^
        | |
        3 3
        """

        digitsToChar = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "qprs",
            "8": "tuv",
            "9": "wxyz"
        }

        def dfs(i, currComb, comb):
            if len(digits) == len(currComb):
                comb.append("".join(currComb))
                return
            if i >= len(digits):
                return

            for c in digitsToChar[digits[i]]:
                #include i
                currComb.append(c)
                dfs(i + 1, currComb, comb)
                currComb.pop()

                # #don't include i
                dfs(i + 1, currComb, comb)

        res = []
        dfs(0, [], res)
        return res

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.swift
class Solution {
    func letterCombinations(_ digits: String) -> [String] {
        let digitsToChar: [Character: [Character]] = [
            "2": Array("abc"),
            "3": Array("def"),
            "4": Array("ghi"),
            "5": Array("jkl"),
            "6": Array("mno"),
            "7": Array("qprs"),
            "8": Array("tuv"),
            "9": Array("wxyz")
        ]

        var comb: [String] = []
        var currComb: [Character] = []
        let digitsArray = Array(digits)

        func dfs(_ i: Int) {
            if digits.count == currComb.count {
                comb.append(String(currComb))
                return
            }
            if i >= digits.count {
                return
            }

            let digit = digitsArray[i]
            guard let letters = digitsToChar[digit] else { return }
            for c in letters {
                //include i
                currComb.append(c)
                dfs(i + 1)
                currComb.removeLast()

                //don't include i
                dfs(i + 1)
            }
        }

        dfs(0)
        return comb
    }
}